phoenix-core/src/main/java/org/apache/phoenix/schema/MetaDataClient.java

```diff
@@ -4168,4 +4176,197 @@ public MutationState useSchema(UseSchemaStatement useSchemaStatement) throws SQL
+                }
+                revokePermissionsFromSchema(hConnection, revokeStatement);
+
+            } else if (revokeStatement.getTableName() != null) {
+                PTable inputTable = PhoenixRuntime.getTable(connection,
+                        SchemaUtil.normalizeFullTableName(revokeStatement.getTableName().toString()));
+                if (!(PTableType.TABLE.equals(inputTable.getType()) || PTableType.SYSTEM.equals(inputTable.getType()))) {
+                    throw new AccessDeniedException("Cannot REVOKE permissions from INDEX TABLES or VIEWS");
+                }
+                revokePermissionsFromTables(hConnection, revokeStatement, inputTable);
+
+            } else {
+                revokePermissionsFromUser(hConnection, revokeStatement);
+            }
+
+        } catch (SQLException e) {
+            // Bubble up the SQL Exception
+            throw e;
+        } catch (Throwable throwable) {
+            // Wrap around other exceptions to PhoenixIOException (Ex: org.apache.hadoop.hbase.security.AccessDeniedException)
+            throw ServerUtil.parseServerException(throwable);
+        }
+
+        return new MutationState(0, 0, connection);
+    }
+
+
+    private void revokePermissionsFromTables(HConnection hConnection, RevokeStatement revokeStatement, PTable inputTable) throws Throwable {
+
+        org.apache.hadoop.hbase.TableName tableName = SchemaUtil.getPhysicalTableName
+                (inputTable.getName().getBytes(), inputTable.isNamespaceMapped());
+
+        revokePermissionsFromTable(hConnection, revokeStatement, tableName);
+
+        for(PTable indexTable : inputTable.getIndexes()) {
+            // Local Indexes don't correspond to new physical table, they are just stored in separate CF of base table.
+            if(indexTable.getIndexType().equals(IndexType.LOCAL)) {
+                continue;
+            }
+            logger.info("Revoking perms from IndexTable: " + indexTable.getName() + " BaseTable: " + inputTable.getName());
+            if (inputTable.isNamespaceMapped() != indexTable.isNamespaceMapped()) {
+                throw new TablesNotInSyncException(inputTable.getTableName().getString(),
+                        indexTable.getTableName().getString(), "Namespace properties");
+            }
+            tableName = SchemaUtil.getPhysicalTableName(indexTable.getName().getBytes(), indexTable.isNamespaceMapped());
+            revokePermissionsFromTable(hConnection, revokeStatement, tableName);
+        }
+
+        byte[] viewIndexTableBytes = MetaDataUtil.getViewIndexPhysicalName(inputTable.getPhysicalName().getBytes());
+        tableName = org.apache.hadoop.hbase.TableName.valueOf(viewIndexTableBytes);
+        boolean viewIndexTableExists = connection.getQueryServices().getAdmin().tableExists(tableName);
+        if(!viewIndexTableExists && inputTable.isMultiTenant()) {
+            logger.error("View Index Table not found for MultiTenant Table: " + inputTable.getName());
```

Comment:
Not sure if this check is needed as we always create the view index table for multi-tenant tables.

Reply:
Theoretically yes, This is more of a sanity check. I don't know if you have ever encountered cases where the index table failed to get created. Since Phoenix operations are not atomic, we can end up in a partial state. The HBase Perms API doesn't check if table exists or not, so it will put up an entry even if the table doesn't exist. Let me know your thoughts and we can decide.